// scroll-into-view-if-needed/index.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _compute = _interopRequireDefault(require("./compute"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var supportsScrollBehavior;

var isFunction = function isFunction(arg) {
  return typeof arg == 'function';
};

var isOptionsObject = function isOptionsObject(options) {
  return options === Object(options) && Object.keys(options).length !== 0;
};

var defaultBehavior = function defaultBehavior(actions, behavior) {
  if (behavior === void 0) {
    behavior = 'auto';
  }

  if (supportsScrollBehavior === undefined) {
    supportsScrollBehavior = 'scrollBehavior' in document.documentElement.style;
  }

  actions.forEach(function (_ref) {
    var el = _ref.el,
        top = _ref.top,
        left = _ref.left;

    if (el.scroll && supportsScrollBehavior) {
      el.scroll({
        top: top,
        left: left,
        behavior: behavior
      });
    } else {
      if (el === document.documentElement) {
        window.scrollTo(left, top);
      } else {
        el.scrollTop = top;
        el.scrollLeft = left;
      }
    }
  });
};

var getOptions = function getOptions(options) {
  if (options === void 0) {
    options = true;
  }

  if (options === true || options === null) {
    return {
      block: 'start',
      inline: 'nearest'
    };
  } else if (options === false) {
    return {
      block: 'end',
      inline: 'nearest'
    };
  } else if (isOptionsObject(options)) {
    return options;
  }

  return {
    block: 'start',
    inline: 'nearest'
  };
};

function scrollIntoView(target, options) {
  if (options === void 0) {
    options = true;
  }

  if (isOptionsObject(options) && isFunction(options.behavior)) {
    return options.behavior((0, _compute.default)(target, options));
  }

  var computeOptions = getOptions(options);
  return defaultBehavior((0, _compute.default)(target, computeOptions), computeOptions.behavior);
}

var _default = scrollIntoView;
exports.default = _default;
module.exports = exports["default"];
